cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -348,9 +348,9 @@ class DatasetWriterDirectoryQueue {
     util::AsyncTaskScheduler::Throttle* throttle_view = throttle.get();
     util::AsyncTaskScheduler* file_scheduler = scheduler_->MakeSubScheduler(
         FileFinishTask{this, std::move(file_queue), std::move(throttle)}, throttle_view);
-    if (init_task_) {
-      file_scheduler->AddSimpleTask(init_task_);
-      init_task_ = {};
+    if (init_future_.is_valid()) {
```

Comment:
What happens if this is not valid?

Reply:
This means that `(directory_.empty() || !write_options_.create_dir)` is `true`.  So either we are not doing a partitioned write (and so don't need to create any directories) or we expect all the directories to be pre-created (and open file will fail later if they are not).

Reply:
Then, because we don't add the task to wait for any "directory creation future" the file scheduler is free to start trying to write to the file immediately.

Reply:
Ah I see. Thanks for the explanation.